beater/middleware/monitoring_middleware.go

```diff
@@ -19,35 +19,62 @@ package middleware
 
 import (
 	"net/http"
+	"sync"
 
 	"github.com/elastic/beats/libbeat/monitoring"
 
 	"github.com/elastic/apm-server/beater/request"
 )
 
+var (
+	mu sync.Mutex
+)
+
+type monitoringMapper struct {
+	resultIDToCounter map[request.ResultID]*monitoring.Int
+	registry          *monitoring.Registry
+}
+
 // MonitoringMiddleware returns a middleware that increases monitoring counters for collecting metrics
 // about request processing. As input parameter it takes a map capable of mapping a request.ResultID to a counter.
-func MonitoringMiddleware(m map[request.ResultID]*monitoring.Int) Middleware {
+func MonitoringMiddleware(registry *monitoring.Registry) Middleware {
 	return func(h request.Handler) (request.Handler, error) {
-		inc := func(id request.ResultID) {
-			if counter, ok := m[id]; ok {
-				counter.Inc()
-			}
-		}
+		mapper := monitoringMapper{resultIDToCounter: map[request.ResultID]*monitoring.Int{}, registry: registry}
+
 		return func(c *request.Context) {
-			inc(request.IDRequestCount)
+			mapper.inc(request.IDRequestCount)
 
 			h(c)
 
-			inc(request.IDResponseCount)
+			mapper.inc(request.IDResponseCount)
 			if c.Result.StatusCode >= http.StatusBadRequest {
-				inc(request.IDResponseErrorsCount)
+				mapper.inc(request.IDResponseErrorsCount)
 			} else {
-				inc(request.IDResponseValidCount)
+				mapper.inc(request.IDResponseValidCount)
 			}
 
-			inc(c.Result.ID)
+			mapper.inc(c.Result.ID)
 		}, nil
+	}
+}
+
+func (m *monitoringMapper) inc(id request.ResultID) {
+	m.mapID(id).Inc()
+}
 
+func (m *monitoringMapper) mapID(id request.ResultID) *monitoring.Int {
+	if ct, ok := m.resultIDToCounter[id]; ok {
+		return ct
 	}
+
+	mu.Lock()
+	defer mu.Unlock()
+	if ct, ok := m.registry.Get(string(id)).(*monitoring.Int); ok {
+		m.resultIDToCounter[id] = ct
+		return ct
```

Comment:
This would be the case when multiple middlewares use monitoring for the same registry and the counter has been already created (they are global) but there is no reference to it from this map. E.g. Intake for RUM and backend use the same `registry`, but they are two different routes with two middlewares

Reply:
wow, that is subtle... it could do with a comment or something. same for the mutex